In [2]:
import pandas as pd

mothers   = pd.read_csv("../data/processed/mothers.csv")
child     = pd.read_csv("../data/processed/child_immunization.csv")
child_mom = pd.read_csv("../data/processed/child_immunization_with_mothers.csv")

mothers.head(), mothers.shape


(   cluster  household  woman_line           caseid  age  urban_rural  region  \
 0        1          8           2      00010008 02   26            1       1   
 1        1         14           2      00010014 02   28            1       1   
 2        1         14           4      00010014 04   17            1       1   
 3        1         21           2      00010021 02   26            1       1   
 4        1         28           3      00010028 03   21            1       1   
 
    education  wealth_quintile  births_5yrs  
 0          2                5            1  
 1          3                5            1  
 2          3                5            0  
 3          2                5            1  
 4          0                5            0  ,
 (8885, 10))

In [3]:
mothers.info()
mothers.describe(include="all")


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8885 entries, 0 to 8884
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   cluster          8885 non-null   int64 
 1   household        8885 non-null   int64 
 2   woman_line       8885 non-null   int64 
 3   caseid           8885 non-null   object
 4   age              8885 non-null   int64 
 5   urban_rural      8885 non-null   int64 
 6   region           8885 non-null   int64 
 7   education        8885 non-null   int64 
 8   wealth_quintile  8885 non-null   int64 
 9   births_5yrs      8885 non-null   int64 
dtypes: int64(9), object(1)
memory usage: 694.3+ KB


cluster    household   woman_line           caseid          age  \
count   8885.000000  8885.000000  8885.000000             8885  8885.000000   
unique          NaN          NaN          NaN             8885          NaN   
top             NaN          NaN          NaN      03050177 02          NaN   
freq            NaN          NaN          NaN                1          NaN   
mean     157.357119    70.921778     2.510748              NaN    27.507597   
std       88.848199    47.133167     1.537766              NaN     9.059677   
min        1.000000     1.000000     1.000000              NaN    15.000000   
25%       82.000000    32.000000     2.000000              NaN    20.000000   
50%      160.000000    65.000000     2.000000              NaN    26.000000   
75%      235.000000   102.000000     3.000000              NaN    35.000000   
max      305.000000   281.000000    20.000000              NaN    49.000000   

        urban_rural       region    education  wealth_quintile  births_5yrs  
count   8885.000000  8885.000000  8885.000000      8885.000000  8885.000000  
unique          NaN          NaN          NaN              NaN          NaN  
top             NaN          NaN          NaN              NaN          NaN  
freq            NaN          NaN          NaN              NaN          NaN  
mean       1.667867     6.029038     0.888689         3.196173     0.647496  
std        0.471005     3.113284     0.929948         1.576186     0.829170  
min        1.000000     1.000000     0.000000         1.000000     0.000000  
25%        1.000000     3.000000     0.000000         2.000000     0.000000  
50%        2.000000     6.000000     1.000000         3.000000     0.000000  
75%        2.000000     9.000000     1.000000         5.000000     1.000000  
max        2.000000    11.000000     3.000000         5.000000     5.000000

In [4]:
child.info()
child.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5753 entries, 0 to 5752
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   cluster     5753 non-null   int64  
 1   household   5753 non-null   int64  
 2   woman_line  5753 non-null   int64  
 3   caseid      5753 non-null   object 
 4   bidx        5753 non-null   int64  
 5   dob_cmc     5753 non-null   int64  
 6   sex         5753 non-null   int64  
 7   bcg         3208 non-null   float64
 8   dpt1        3208 non-null   float64
 9   dpt2        3208 non-null   float64
 10  dpt3        3208 non-null   float64
 11  polio0      3208 non-null   float64
 12  polio1      3208 non-null   float64
 13  polio2      3208 non-null   float64
 14  polio3      3208 non-null   float64
 15  measles1    3208 non-null   float64
dtypes: float64(9), int64(6), object(1)
memory usage: 719.2+ KB


cluster    household   woman_line         bidx      dob_cmc  \
count  5753.000000  5753.000000  5753.000000  5753.000000  5753.000000   
mean    145.726577    68.043456     2.059447     1.354598  2008.679645   
std      84.922632    47.054656     1.149444     0.574526     1.464021   
min       1.000000     1.000000     1.000000     1.000000  2006.000000   
25%      75.000000    29.000000     2.000000     1.000000  2007.000000   
50%     141.000000    62.000000     2.000000     1.000000  2009.000000   
75%     213.000000    98.000000     2.000000     2.000000  2010.000000   
max     305.000000   273.000000    20.000000     5.000000  2011.000000   

               sex          bcg         dpt1         dpt2         dpt3  \
count  5753.000000  3208.000000  3208.000000  3208.000000  3208.000000   
mean      1.483921     1.551434     0.825125     0.858479     0.877494   
std       0.499785     1.583111     0.933207     1.051785     0.861622   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%       1.000000     0.000000     0.000000     0.000000     0.000000   
50%       1.000000     1.000000     1.000000     1.000000     1.000000   
75%       2.000000     4.000000     1.000000     1.000000     1.000000   
max       2.000000     4.000000     8.000000     8.000000     8.000000   

            polio0       polio1       polio2       polio3     measles1  
count  3208.000000  3208.000000  3208.000000  3208.000000  3208.000000  
mean      0.779925     0.744701     0.640274     0.560474     0.607232  
std       1.067802     0.858551     1.064915     0.825284     1.028910  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       1.000000     1.000000     0.000000     0.000000     0.000000  
75%       1.000000     1.000000     1.000000     1.000000     1.000000  
max       8.000000     8.000000     8.000000     8.000000     8.000000

In [5]:
child_mom.info()
child_mom.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5753 entries, 0 to 5752
Data columns (total 23 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   cluster          5753 non-null   int64  
 1   household        5753 non-null   int64  
 2   woman_line       5753 non-null   int64  
 3   caseid           5753 non-null   object 
 4   bidx             5753 non-null   int64  
 5   dob_cmc          5753 non-null   int64  
 6   sex              5753 non-null   int64  
 7   bcg              3208 non-null   float64
 8   dpt1             3208 non-null   float64
 9   dpt2             3208 non-null   float64
 10  dpt3             3208 non-null   float64
 11  polio0           3208 non-null   float64
 12  polio1           3208 non-null   float64
 13  polio2           3208 non-null   float64
 14  polio3           3208 non-null   float64
 15  measles1         3208 non-null   float64
 16  caseid_mother    5753 non-null   object 
 17  age           

cluster  household  woman_line           caseid  bidx  dob_cmc  sex  bcg  \
0        1          8           2      00010008 02     1     2007    2  NaN   
1        1         14           2      00010014 02     1     2007    2  NaN   
2        1         21           2      00010021 02     1     2011    2  1.0   
3        1         41           2      00010041 02     1     2010    1  NaN   
4        1         47           2      00010047 02     1     2009    2  1.0   

   dpt1  dpt2  ...  polio2  polio3  measles1    caseid_mother  age  \
0   NaN   NaN  ...     NaN     NaN       NaN      00010008 02   26   
1   NaN   NaN  ...     NaN     NaN       NaN      00010014 02   28   
2   1.0   1.0  ...     0.0     0.0       0.0      00010021 02   26   
3   NaN   NaN  ...     NaN     NaN       NaN      00010041 02   30   
4   1.0   1.0  ...     1.0     1.0       1.0      00010047 02   28   

   urban_rural region  education  wealth_quintile  births_5yrs  
0            1      1          2                5            1  
1            1      1          3                5            1  
2            1      1          2                5            1  
3            1      1          3                5            1  
4            1      1          2                5            1  

[5 rows x 23 columns]

In [6]:
import pandas as pd

mothers = pd.read_csv("../data/processed/mothers.csv")
child = pd.read_csv("../data/processed/child_immunization.csv")
child_mom = pd.read_csv("../data/processed/child_immunization_with_mothers.csv")

vaccine_cols = ["bcg", "dpt1", "dpt2", "dpt3", "polio0", "polio1", "polio2", "polio3", "measles1"]

# how many doses each child has received
child_mom["n_received"] = child_mom[vaccine_cols].sum(axis=1)

# target: 1 = missed at least one recommended vaccine, 0 = got all
child_mom["missed_any"] = (child_mom[vaccine_cols] < 1).any(axis=1).astype(int)

child_mom[["n_received", "missed_any"] + vaccine_cols].head()


n_received  missed_any  bcg  dpt1  dpt2  dpt3  polio0  polio1  polio2  \
0         0.0           0  NaN   NaN   NaN   NaN     NaN     NaN     NaN   
1         0.0           0  NaN   NaN   NaN   NaN     NaN     NaN     NaN   
2         6.0           1  1.0   1.0   1.0   1.0     1.0     1.0     0.0   
3         0.0           0  NaN   NaN   NaN   NaN     NaN     NaN     NaN   
4         9.0           0  1.0   1.0   1.0   1.0     1.0     1.0     1.0   

   polio3  measles1  
0     NaN       NaN  
1     NaN       NaN  
2     0.0       0.0  
3     NaN       NaN  
4     1.0       1.0

In [7]:
# Overall rate of missed vaccination
child_mom["missed_any"].value_counts(normalize=True)

# Missed rate by key factors
def missed_rate(group_col):
    return (
        child_mom.groupby(group_col)["missed_any"]
        .mean()
        .sort_values(ascending=False)
    )

missed_rate("region")
missed_rate("urban_rural")
missed_rate("education")
missed_rate("wealth_quintile")
missed_rate("births_5yrs")


births_5yrs
3    0.432217
2    0.406130
1    0.367089
4    0.326923
5    0.266667
Name: missed_any, dtype: float64

In [8]:
model_cols = [
    "cluster", "household", "woman_line",
    "dob_cmc", "sex",
    "age", "urban_rural", "region",
    "education", "wealth_quintile", "births_5yrs",
] + vaccine_cols + ["n_received", "missed_any"]

model_df = child_mom[model_cols].copy()
model_df.to_csv("../data/processed/child_mom_model_table.csv", index=False)
model_df.shape


(5753, 22)

In [10]:
# Class balance
child_mom["missed_any"].value_counts(normalize=True)


missed_any
0    0.608726
1    0.391274
Name: proportion, dtype: float64

In [11]:
child_mom.groupby("region")["missed_any"].mean().sort_values(ascending=False)


region
2     0.527607
5     0.456829
4     0.424200
7     0.415152
8     0.406667
9     0.402685
6     0.333962
11    0.330846
3     0.328767
1     0.292952
10    0.216495
Name: missed_any, dtype: float64

In [12]:
model_df.to_csv("../data/processed/child_mom_model_table.csv", index=False)


In [15]:
model_df.columns.tolist()


['cluster',
 'household',
 'woman_line',
 'dob_cmc',
 'sex',
 'age',
 'urban_rural',
 'region',
 'education',
 'wealth_quintile',
 'births_5yrs',
 'bcg',
 'dpt1',
 'dpt2',
 'dpt3',
 'polio0',
 'polio1',
 'polio2',
 'polio3',
 'measles1',
 'n_received',
 'missed_any']

In [16]:
target_col = "missed_any"

# ID-like columns that we DON'T want as features
id_cols = ["cluster", "household", "woman_line"]   # no "caseid" here

# 1) Features matrix: drop target + ID columns
X = model_df.drop(columns=id_cols + [target_col])

# 2) Target vector
y = model_df[target_col].astype(int)

# 3) Handle missing values (simple baseline: treat NaNs as 0)
X = X.fillna(0)

X.head(), X.shape, X.dtypes


(   dob_cmc  sex  age  urban_rural  region  education  wealth_quintile  \
 0     2007    2   26            1       1          2                5   
 1     2007    2   28            1       1          3                5   
 2     2011    2   26            1       1          2                5   
 3     2010    1   30            1       1          3                5   
 4     2009    2   28            1       1          2                5   
 
    births_5yrs  bcg  dpt1  dpt2  dpt3  polio0  polio1  polio2  polio3  \
 0            1  0.0   0.0   0.0   0.0     0.0     0.0     0.0     0.0   
 1            1  0.0   0.0   0.0   0.0     0.0     0.0     0.0     0.0   
 2            1  1.0   1.0   1.0   1.0     1.0     1.0     0.0     0.0   
 3            1  0.0   0.0   0.0   0.0     0.0     0.0     0.0     0.0   
 4            1  1.0   1.0   1.0   1.0     1.0     1.0     1.0     1.0   
 
    measles1  n_received  
 0       0.0         0.0  
 1       0.0         0.0  
 2       0.0         6.0  


In [17]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, roc_auc_score

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

clf = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)
y_proba = clf.predict_proba(X_test)[:, 1]

print(classification_report(y_test, y_pred))
print("ROC AUC:", roc_auc_score(y_test, y_proba))


              precision    recall  f1-score   support

           0       0.77      0.98      0.86       701
           1       0.95      0.55      0.69       450

    accuracy                           0.81      1151
   macro avg       0.86      0.76      0.78      1151
weighted avg       0.84      0.81      0.80      1151

ROC AUC: 0.8460722776985259


c:\Users\User\miniconda3\lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [19]:
model_df.isna().sum()


cluster               0
household             0
woman_line            0
dob_cmc               0
sex                   0
age                   0
urban_rural           0
region                0
education             0
wealth_quintile       0
births_5yrs           0
bcg                2545
dpt1               2545
dpt2               2545
dpt3               2545
polio0             2545
polio1             2545
polio2             2545
polio3             2545
measles1           2545
n_received            0
missed_any            0
dtype: int64

In [20]:
vaccine_cols = ["bcg", "dpt1", "dpt2", "dpt3",
                "polio0", "polio1", "polio2", "polio3",
                "measles1"]

# Treat missing vaccine info as "dose not received" (0)
model_df[vaccine_cols] = model_df[vaccine_cols].fillna(0)

# Quick check
model_df[vaccine_cols].isna().sum()


bcg         0
dpt1        0
dpt2        0
dpt3        0
polio0      0
polio1      0
polio2      0
polio3      0
measles1    0
dtype: int64

In [21]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, roc_auc_score, confusion_matrix

target_col = "missed_any"

X = model_df.drop(columns=[target_col])
y = model_df[target_col].astype(int)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

clf = make_pipeline(
    StandardScaler(),
    LogisticRegression(max_iter=5000)  # class_weight='balanced' is an option to try later
)

clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)
y_proba = clf.predict_proba(X_test)[:, 1]

print(classification_report(y_test, y_pred))
print("ROC AUC:", roc_auc_score(y_test, y_proba))
print("Confusion matrix:\n", confusion_matrix(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.90      0.94      0.92       701
           1       0.90      0.83      0.86       450

    accuracy                           0.90      1151
   macro avg       0.90      0.88      0.89      1151
weighted avg       0.90      0.90      0.89      1151

ROC AUC: 0.9485021398002854
Confusion matrix:
 [[658  43]
 [ 77 373]]


In [22]:
model_df[vaccine_cols] = model_df[vaccine_cols].fillna(0)


In [24]:
target_col = "missed_any"

# 1) Features / target
id_cols = ["cluster", "household", "woman_line"]   # <-- no caseid here
X = model_df.drop(columns=id_cols + [target_col])
y = model_df[target_col].astype(int)

# 2) Train / test split (stratified)
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, roc_auc_score, confusion_matrix

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# 3) Pipeline
clf = make_pipeline(
    StandardScaler(),
    LogisticRegression(max_iter=5000)
)

clf.fit(X_train, y_train)

y_pred  = clf.predict(X_test)
y_proba = clf.predict_proba(X_test)[:, 1]

print(classification_report(y_test, y_pred))
print("ROC AUC:", roc_auc_score(y_test, y_proba))
print("Confusion matrix:\n", confusion_matrix(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.90      0.94      0.92       701
           1       0.90      0.83      0.86       450

    accuracy                           0.90      1151
   macro avg       0.90      0.88      0.89      1151
weighted avg       0.90      0.90      0.89      1151

ROC AUC: 0.9484688540180695
Confusion matrix:
 [[658  43]
 [ 77 373]]


Initial predictive model: missed vaccinations

Target variable

- `missed_any`  
  - `1` = child missed at least one recommended vaccine dose  
  - `0` = child received all recommended doses

Features used

- Child & household characteristics:  
  `dob_cmc`, `sex`, `age`, `urban_rural`, `region`, `education`, `wealth_quintile`, `births_5yrs`
- Vaccination history (binary / counts):  
  `bcg`, `dpt1`, `dpt2`, `dpt3`, `polio0`, `polio1`, `polio2`, `polio3`, `measles1`, `n_received`
- ID-like columns (`cluster`, `household`, `woman_line`) were **excluded** from the model.

Data cleaning for modelling

- Merged child and mother tables into `child_mom`
- Built `missed_any` from individual vaccine dose columns  
- All missing vaccine doses in `bcg`–`measles1` were imputed as `0` (interpreted as “dose not received”):
  - `model_df[vaccine_cols] = model_df[vaccine_cols].fillna(0)`
- Train/test split: 80% train, 20% test, stratified by `missed_any`

Model

- Pipeline: `StandardScaler` → `LogisticRegression(max_iter=5000)`
- Reason:
  - Scaling helps optimizer converge (features on comparable scale)
  - Higher `max_iter` avoids convergence warnings

Performance (test set)

- Accuracy: ~0.90  
- ROC AUC: ~0.95  
- Class-wise:
  - Class `0` (no missed doses): precision 0.90, recall 0.94, F1 0.92 (support = 701)
  - Class `1` (missed ≥1 dose): precision 0.90, recall 0.83, F1 0.86 (support = 450)
- Confusion matrix (`[[TN FP], [FN TP]]`):
  - `[[658  43], [77 373]]`
  - Model correctly identifies most children both with and without missed doses, with recall for the “missed” class reasonably high (0.83).

Takeaway

A simple logistic regression model with basic socio-demographic and vaccine history features already achieves strong discrimination (ROC AUC ≈ 0.95). This is a solid baseline for predicting which children are at risk of missing vaccinations and can be compared to more complex models later (e.g. tree-based methods).
